tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/LetInTypeSynchronizer.scala

```diff
@@ -0,0 +1,75 @@
+package at.forsyte.apalache.tla.pp
+
+import at.forsyte.apalache.tla.lir.TypedPredefs.TypeTagAsTlaType1
+import at.forsyte.apalache.tla.lir.{LetInEx, OperEx, TlaDecl, TlaEx, TlaOperDecl, Typed, TypingException}
+import at.forsyte.apalache.tla.lir.transformations.{TlaExTransformation, TransformationTracker}
+import at.forsyte.apalache.tla.typecheck.etc.{Substitution, TypeUnifier}
+
+/**
+ * LetInTypeSynchronizer re-computes the types of any residual LET-IN defined operators, that were defined
+ * in the scope of a polymorphic operator (e.g. in FoldX). After inlining, any leftover LET-IN operators should have monotypes
+ */
+class LetInTypeSynchronizer(tracker: TransformationTracker) extends TlaExTransformation {
+  override def apply(ex: TlaEx): TlaEx = transform(ex)
+
+  private def isRelevantLetInEx(letInEx: LetInEx): Boolean = {
+    // currently, we only care about let-ins embedded in fold, which are nonnullary and single-def
+    letInEx.decls.length == 1 && letInEx.decls.forall(_.formalParams.nonEmpty)
+  }
+
+  def transform: TlaExTransformation = tracker.trackEx {
+    // interesting case
+    case letInEx @ LetInEx(letInBody, defs @ _*) if isRelevantLetInEx(letInEx) =>
+      // The problem with let-ins is that, while the body will have the correct monotype, the
+      // type checking algorithm does not update the declaration and body tags.
+
+      // The solution: Unify the body type and the declaration type, then apply the resulting substitution over the
+      // declaration body
+      val Seq(letInDecl) = defs
+      val bodyType = letInBody.typeTag.asTlaType1()
+      val declType = letInDecl.typeTag.asTlaType1()
+      val unifiedOpt = (new TypeUnifier).unify(Substitution.empty, bodyType, declType)
+
+      unifiedOpt match {
+        case None => throw new TypingException("Could not synchronize type of fold operator.")
+        case Some((subst, unifiedType)) =>
+          val substitutor = new TypeSubstitutor(tracker, subst)
+          val newDecl = letInDecl.copy(body = substitutor(letInDecl.body))(Typed(unifiedType))
+          LetInEx(letInBody, newDecl)(Typed(unifiedType))
+      }
+
+    case letInEx @ LetInEx(letInBody, defs @ _*) =>
+      // assume !isRelevant
```

Comment:
what is happening here? Could you add a comment?

Reply:
What part of this is unclear/needs comments?

Reply:
This part. What is `!isRelevant`? What happens in the code below?

Reply:
https://github.com/informalsystems/apalache/blob/e3842c0ce295b87b70e5b3d67874430740aa4eeb/tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/LetInTypeSynchronizer.scala#L15-L22

Reply:
I do not understand, why we should spend 1 hour firing messages, instead of writing two lines of comments?